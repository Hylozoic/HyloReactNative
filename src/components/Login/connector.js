import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { login, loginWithFacebook, loginWithGoogle } from './actions'

function mapStateToProps (state) {
  return {
    error: state.session.loginError,
    defaultEmail: state.session.defaultLoginEmail
  }
}

function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators({
      login,
      loginWithFacebook,
      loginWithGoogle
    }, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
