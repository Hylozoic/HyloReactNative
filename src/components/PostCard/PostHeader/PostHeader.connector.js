import { connect } from 'react-redux'
import getMe from '../../../store/selectors/getMe'
import getCommunity from '../../../store/selectors/getCommunity'

export function mapStateToProps (state, props) {
  const community = getCommunity(state, props)
  return {
    currentUser: getMe(state, props),
    community
  }
}

export function mapDispatchToProps (dispatch, props) {
  return {
  }
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  const { currentUser, community } = stateProps
  const { creator, editPost } = ownProps
  const canEdit = currentUser && creator && currentUser.id === creator.id
  const canFlag = currentUser && creator && currentUser.id !== creator.id
  const canModerate = currentUser && currentUser.canModerate(community)
  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    editPost: canEdit ? () => editPost() : null,
    canEdit,
    canFlag
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
