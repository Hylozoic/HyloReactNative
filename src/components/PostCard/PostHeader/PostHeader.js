import React from 'react'
import { View, Text, TouchableOpacity, Alert, FlatList } from 'react-native'
import { get, filter, isEmpty } from 'lodash/fp'
import { humanDate } from 'hylo-utils/text'
import {
  rhino30,
  rhino50,
  caribbeanGreen,
  prim,
  fuchsiaPink,
  mangoYellow,
  havelockBlue,
  westSide,
  jade,
  fakeAlpha
} from 'style/colors'
import Avatar from 'components/Avatar'
import Icon from 'components/Icon'
import PopupMenuButton from 'components/PopupMenuButton'
import FlagContent from 'components/FlagContent'

export default class PostHeader extends React.PureComponent {
  constructor (props) {
    super(props)
    this.state = {
      creator: {},
      flaggingVisible: false
    }
  }

  handleEditPost = () => {
    this.props.editPost(this.props.postId)
  }

  handleShowMember = () => {
    this.props.showMember && this.props.showMember(this.props.creator.id)
  }

  topicKeyExtractor = item => item.id

  renderTopic = ({ item }) => (
    <TouchableOpacity onPress={() => this.props.showTopic && this.props.showTopic(item.name)}>
      <Text style={styles.topicLabel}>#{item.name}</Text>
    </TouchableOpacity>
  )

  render () {
    const {
      creator: { avatarUrl, name, tagline },
      date,
      type,
      postId,
      slug,
      canFlag,
      removePost,
      deletePost,
      deletePostAndClose,
      closeOnDelete,
      pinned,
      pinPost,
      topics,
      announcement,
      canEdit,
      hideMenu,
      hideDateRow,
      smallAvatar
    } = this.props
    const { flaggingVisible } = this.state
    const showTopics = !isEmpty(topics)
    // Used to generate a link to this post from the backend.
    const linkData = {
      slug,
      id: postId,
      type: 'post'
    }

    let flagPost
    if (canFlag) {
      flagPost = () => {
        this.setState({ flaggingVisible: true })
      }
    }

    const deletePostWithConfirm = deletePost
      ? () => Alert.alert(
          'Confirm Delete',
          'Are you sure you want to delete this post?',
          [
            {
              text: 'Yes',
              onPress: () => closeOnDelete
                ? deletePostAndClose()
                : deletePost()
            },
            { text: 'Cancel', style: 'cancel' }
          ]
        )
      : null

    const removePostWithConfirm = removePost
      ? () => Alert.alert(
          'Confirm Removal',
          'Are you sure you want to remove this post from this group?',
          [
            { text: 'Yes', onPress: () => removePost() },
            { text: 'Cancel', style: 'cancel' }
          ])
      : null

    return (
      <View style={styles.container}>
        <View style={styles.avatarSpacing}>
          <TouchableOpacity onPress={this.handleShowMember}>
            {!!avatarUrl && <Avatar avatarUrl={avatarUrl} dimension={smallAvatar && 20} />}
          </TouchableOpacity>
        </View>
        <View style={styles.meta}>
          <TouchableOpacity onPress={this.handleShowMember}>
            {name && <Text style={styles.username}>{name}</Text>}
            {!!tagline && <Text style={styles.metaText}>{tagline}</Text>}
          </TouchableOpacity>
          {!hideDateRow && (
            <View style={styles.dateRow}>
              <Text style={styles.metaText}>{humanDate(date)}</Text>
              {!!showTopics && (
                <FlatList
                  data={topics}
                  style={styles.topicList}
                  horizontal
                  keyExtractor={this.topicKeyExtractor}
                  showsHorizontalScrollIndicator={false}
                  renderItem={this.renderTopic}
                />
              )}
            </View>
          )}
        </View>
        <View style={styles.upperRight}>
          {pinned && <Icon name='Pin' style={styles.pinIcon} />}
          {announcement && <Icon name='Announcement' style={styles.announcementIcon} />}
          {type && <PostLabel type={type} />}
          {!hideMenu && (
            <PostMenu
              removePost={removePostWithConfirm}
              deletePost={deletePostWithConfirm}
              editPost={canEdit && this.handleEditPost}
              flagPost={flagPost}
              pinPost={pinPost}
              pinned={pinned}
            />
          )}
          {flaggingVisible && (
            <FlagContent
              type='post'
              linkData={linkData}
              onClose={() => this.setState({ flaggingVisible: false })}
            />
          )}
        </View>
      </View>
    )
  }
}

export function PostMenu ({ deletePost, editPost, flagPost, removePost, pinPost, pinned }) {
  // If the function is defined, than it's a valid action
  const flagLabel = 'Flag This Post'
  const deleteLabel = 'Delete This Post'
  const removeLabel = 'Remove Post From Group'
  const pinLabel = pinned ? 'Unpin' : 'Pin'

  const actions = filter(x => x[1], [
    [deleteLabel, deletePost],
    [flagLabel, flagPost],
    [removeLabel, removePost],
    ['Edit This Post', editPost],
    [pinLabel, pinPost]
  ])

  if (isEmpty(actions)) return null

  const destructiveLabels = [flagLabel, deleteLabel, removeLabel]
  const destructiveButtonIndex = destructiveLabels.includes(actions[0][0]) ? 0 : -1

  return (
    <PopupMenuButton
      actions={actions}
      hitSlop={{ top: 20, bottom: 10, left: 10, right: 15 }}
      destructiveButtonIndex={destructiveButtonIndex}
    >
      <Icon name='More' style={styles.menuIcon} />
    </PopupMenuButton>
  )
}

export function PostLabel ({ type }) {
  // Prevent redboxing on missing type in styles object
  // TODO: is this the default style we want to use?
  const labelTypeStyle = get(type, labelStyles) || labelStyles.discussion
  const boxStyle = [labelStyles.box, labelTypeStyle.box]
  const textStyle = [labelStyles.text, labelTypeStyle.text]
  return (
    <View style={boxStyle}>
      <Text style={textStyle}>
        {type.toUpperCase()}
      </Text>
    </View>
  )
}

const styles = {
  container: {
    flexDirection: 'row',
    marginBottom: 10
  },
  meta: {
    paddingTop: 7 + 9,
    paddingRight: 7,
    flex: 1,
    justifyContent: 'space-around'
  },
  avatarSpacing: {
    paddingTop: 7 + 9,
    paddingLeft: 7 + 5,
    paddingRight: 7
  },
  avatarNudge: {
    paddingTop: 18
  },
  username: {
    fontSize: 14,
    color: '#363D3C',
    fontFamily: 'Circular-Bold'
  },
  usernameNudge: {
    top: 10
  },
  dateRow: {
    flexDirection: 'row',
    flex: 1
  },
  metaText: {
    fontSize: 12,
    color: rhino30,
    fontFamily: 'Circular-Book'
  },
  spacer: {
    fontSize: 12,
    color: rhino30,
    fontFamily: 'Circular-Book',
    marginHorizontal: 5
  },
  contextLabel: {
    fontSize: 12,
    fontFamily: 'Circular-Book',
    color: caribbeanGreen
  },
  topicList: {
    marginLeft: 4,
    flex: 1
  },
  topicLabel: {
    fontSize: 12,
    paddingRight: 5,
    flex: 1,
    fontFamily: 'Circular-Book',
    color: caribbeanGreen
  },
  upperRight: {
    position: 'absolute',
    paddingTop: 14,
    paddingRight: 6,
    right: 0,
    flexDirection: 'row',
    alignItems: 'flex-start'
  },
  pinIcon: {
    fontSize: 20,
    color: rhino50,
    marginRight: 10
  },
  menuIcon: {
    fontSize: 20,
    paddingLeft: 10,
    paddingRight: 5,
    color: rhino50
  },
  announcementIcon: {
    color: caribbeanGreen,
    fontSize: 20,
    alignItems: 'flex-end',
    marginRight: 2
  }
}

const labelStyles = {
  box: {
    borderRadius: 4,
    paddingTop: 3,
    paddingBottom: 3,
    paddingLeft: 6,
    paddingRight: 6,
    marginRight: 4,
    marginTop: 1
  },
  text: {
    fontSize: 10,
    fontFamily: 'Circular-Bold',
    letterSpacing: 0.8
  },
  discussion: {
    box: {
      backgroundColor: fakeAlpha(havelockBlue, 0.2)
    },
    text: {
      color: havelockBlue
    }
  },
  event: {
    box: {
      backgroundColor: 'rgba(254, 72, 80, .2)'
    },
    text: {
      color: 'rgba(254, 72, 80, 1)'
    }
  },
  offer: {
    box: {
      backgroundColor: fakeAlpha(jade, 0.2)
    },
    text: {
      color: jade
    }
  },
  resource: {
    box: {
      backgroundColor: fakeAlpha(mangoYellow, 0.2)
    },
    text: {
      color: mangoYellow
    }
  },
  project: {
    box: {
      backgroundColor: fakeAlpha(westSide, 0.2)
    },
    text: {
      color: westSide
    }
  },
  request: {
    box: {
      backgroundColor: prim
    },
    text: {
      color: fuchsiaPink
    }
  }
}
