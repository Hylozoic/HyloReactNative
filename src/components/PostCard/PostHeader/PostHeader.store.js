export const MODULE_NAME = 'PostHeader'

// Constants
export const DELETE_POST = `${MODULE_NAME}/DELETE_POST`
export const DELETE_POST_PENDING = DELETE_POST + '_PENDING'
export const REMOVE_POST = `${MODULE_NAME}/REMOVE_POST`
export const REMOVE_POST_PENDING = REMOVE_POST + '_PENDING'
export const PIN_POST = `${MODULE_NAME}/PIN_POST`
export const PIN_POST_PENDING = `${PIN_POST}_PENDING`

// Action Creators
export function deletePost (id) {
  return {
    type: DELETE_POST,
    graphql: {
      query: `mutation ($id: ID) {
        deletePost(id: $id) {
          success
        }
      }`,
      variables: {
        id
      }
    },
    meta: {
      optimistic: true,
      id
    }
  }
}

export function removePost (postId, slug) {
  return {
    type: REMOVE_POST,
    graphql: {
      query: `mutation ($postId: ID, $slug: String) {
        removePost(postId: $postId, slug: $slug) {
          success
        }
      }`,
      variables: {
        postId,
        slug
      }
    },
    meta: {
      optimistic: true,
      postId,
      slug
    }
  }
}

export function pinPost (postId, groupId) {
  return {
    type: PIN_POST,
    graphql: {
      query: `mutation ($postId: ID, $groupId: ID) {
        pinPost(postId: $postId, groupId: $groupId) {
          success
        }
      }`,
      variables: {
        postId,
        groupId
      }
    },
    meta: {
      optimistic: true,
      postId,
      groupId
    }
  }
}

export function ormSessionReducer ({ Post }, { type, meta }) {
  switch (type) {
    case DELETE_POST_PENDING:
      Post.withId(meta.id).delete()
      break

    case REMOVE_POST_PENDING:
      const post = Post.withId(meta.postId)
      const groups = post.groups.filter(c =>
        c.slug !== meta.slug).toModelArray()
      post.update({ groups })
      break
  }
}
