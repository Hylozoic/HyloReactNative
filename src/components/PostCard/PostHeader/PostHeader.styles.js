import { POST_TYPES } from 'store/models/Post'
import { rhino30, rhino50, caribbeanGreen, rhino40, white80onCaribbeanGreen, white40onCaribbeanGreen } from 'style/colors'

export const styles = {
  badgeRow: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'center',
    gap: 1,
    marginRight: 2
  },
  container: {
    flexDirection: 'row',
    marginBottom: 8,
    borderBottomColor: 'rgba(235, 235, 235, 1.0)',
    borderBottomWidth: 1,
    paddingVertical: 9,
    paddingHorizontal: 9,
    alignItems: 'center'
  },
  nameAndDate: {
    paddingRight: 7,
    flex: 1,
    flexDirection: 'row',
    alignItems: 'center'
  },
  name: {
    fontSize: 13,
    color: '#363D3C',
    fontFamily: 'Circular-Bold',
    marginRight: 4
  },
  date: {
    marginLeft: 2,
    fontSize: 12,
    color: rhino40,
    fontFamily: 'Circular-Book'
  },
  avatarSpacing: {
    paddingRight: 7
  },
  spacer: {
    fontSize: 12,
    color: rhino30,
    fontFamily: 'Circular-Book',
    marginHorizontal: 5
  },
  contextLabel: {
    fontSize: 12,
    fontFamily: 'Circular-Book',
    color: caribbeanGreen
  },
  topicList: {
    marginLeft: 4,
    flex: 1
  },
  allBadgesPill:{
    backgroundColor: white40onCaribbeanGreen,
    borderRadius: 30,
    border: 1,
    padding: 2,
    flex: 1,
    flexDirection: 'row',
    borderColor: white40onCaribbeanGreen,
    alignItems: 'center',
    justifyContent: 'center',
    position: 'absolute',
    top: -34
  },
  badgePill: {
    backgroundColor: white80onCaribbeanGreen,
    borderRadius: 30,
    border: 1,
    padding: 2,
    gap: -3,
    flex: 1,
    flexDirection: 'row',
    borderColor: white80onCaribbeanGreen,
    alignItems: 'center',
    justifyContent: 'center'
  },
  topicLabel: {
    fontSize: 12,
    paddingRight: 5,
    flex: 1,
    fontFamily: 'Circular-Book',
    color: caribbeanGreen
  },
  upperRight: {
    flexDirection: 'row',
    alignItems: 'flex-start'
  },
  pinIcon: {
    fontSize: 20,
    color: rhino50,
    marginRight: 10
  },
  moreIcon: {
    fontSize: 20,
    paddingLeft: 5,
    color: rhino50
  },
  announcementIcon: {
    color: caribbeanGreen,
    fontSize: 20,
    alignItems: 'flex-end',
    marginRight: 2
  }
}

export const labelStyles = {
  box: {
    borderRadius: 4,
    paddingTop: 3,
    paddingBottom: 3,
    paddingLeft: 6,
    paddingRight: 6,
    marginRight: 0,
    marginTop: 1
  },
  text: {
    fontSize: 10,
    fontFamily: 'Circular-Bold',
    letterSpacing: 0.8
  },
  discussion: {
    box: {
      backgroundColor: POST_TYPES.discussion.backgroundColor
    },
    text: {
      color: POST_TYPES.discussion.primaryColor
    }
  },
  event: {
    box: {
      backgroundColor: POST_TYPES.event.backgroundColor
    },
    text: {
      color: POST_TYPES.event.primaryColor
    }
  },
  offer: {
    box: {
      backgroundColor: POST_TYPES.offer.backgroundColor
    },
    text: {
      color: POST_TYPES.offer.primaryColor
    }
  },
  resource: {
    box: {
      backgroundColor: POST_TYPES.resource.backgroundColor
    },
    text: {
      color: POST_TYPES.resource.primaryColor
    }
  },
  project: {
    box: {
      backgroundColor: POST_TYPES.project.backgroundColor
    },
    text: {
      color: POST_TYPES.project.primaryColor
    }
  },
  request: {
    box: {
      backgroundColor: POST_TYPES.request.backgroundColor
    },
    text: {
      color: POST_TYPES.request.primaryColor
    }
  }
}

export default styles
