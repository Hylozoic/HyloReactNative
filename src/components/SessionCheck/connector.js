import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { checkSession } from './actions'
import getMe from '../../store/selectors/getMe'
import { get } from 'lodash/fp'

export function mapStateToProps (state) {
  const signupInProgress = get('settings.signupInProgress', getMe(state))

  return {
    loggedIn: state.session.loggedIn && !signupInProgress
  }
}

export function mapDispatchToProps (dispatch) {
  return {
    actions: bindActionCreators({
      checkSession
    }, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)
