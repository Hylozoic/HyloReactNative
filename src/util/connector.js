/**
 * @providesModule util/connector
 */

export const mapWhenFocused = mapper => (stateOrDispatch, props) => props.isFocused
  ? mapper(stateOrDispatch, props)
  : props

export const mergeWhenFocused = merger => (stateProps, dispatchProps, ownProps) =>
  ownProps.isFocused ? merger(stateProps, dispatchProps, ownProps) : ownProps
