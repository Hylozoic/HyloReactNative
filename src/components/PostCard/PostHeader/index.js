import component from './PostHeader'
import connector from './PostHeader.connector'

export default connector(component)
