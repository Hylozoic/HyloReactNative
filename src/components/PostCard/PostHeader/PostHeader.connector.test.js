import { mergeProps } from './PostHeader.connector'

describe('mergeProps', () => {
  it('cannot Flag when user is creator', () => {
    const stateProps = {
      currentUser: {id: 20, canModerate: () => false},
    }
    const ownProps = {creator: {id: 20}}
    const { canFlag } = mergeProps(stateProps, { }, ownProps)

    expect(canFlag).toBeFalsy()
  })

  it('can Flag when user is not creator', () => {
    const stateProps = {
      currentUser: {id: 20, canModerate: () => false}
    }
    const ownProps = {creator: {id: 40}}
    const { canFlag } = mergeProps(stateProps, { }, ownProps)

    expect(canFlag).toBeTruthy()
  })
  
  it('can edit post when user is creator', () => {
    const stateProps = {
      currentUser: {id: 20, canModerate: () => false},
    }
    const ownProps = {creator: {id: 20}}
    const { canEdit, editPost } = mergeProps(stateProps, { }, ownProps)

    expect(canEdit).toBeTruthy()
    expect(editPost).toBeTruthy()
  })

  it('cannot edit post when user is not creator', () => {
    const stateProps = {
      currentUser: {id: 20, canModerate: () => false},
    }
    const ownProps = {creator: {id: 40}}
    const { canEdit, editPost } = mergeProps(stateProps, { }, ownProps)

    expect(canEdit).toBeFalsy()
    expect(editPost).toBeFalsy()
  })
})
