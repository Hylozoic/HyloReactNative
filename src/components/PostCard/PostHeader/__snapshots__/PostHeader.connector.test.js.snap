// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapDispatchToProps returns the right keys 1`] = `
Object {
  "deletePost": [Function],
  "pinPost": [Function],
  "removePost": [Function],
}
`;
